### PR TITLE
remove static/random NPC options, adjust random NPC rates

### PR DESCRIPTION
**Pull request by esotericist**
#### Summary
SUMMARY: Features "remove static/random NPC options, adjust random NPC rates"

#### Purpose of change

With the ongoing changes in the direction of needing NPCs for core game mechanics, it's going to be increasingly problematic if players don't have NPCs available whether that's due to having an old default config file, or disabling out of ignorance of the current design of the game. Further, random NPCs currently show up far less often than expected.

#### Describe the solution

Remove the "Static NPCs", "Starting NPCs Spawn", and "Random NPCs" configuration settings, and all associated conditional code.

Change random NPC density config to "Random NPC spawn time", specified in days instead of an opaque scalar value. Default value set to 4 days (which corresponds to a `1.0` value using the previous config system), and change random NPC spawning code to ignore static/starting NPCs when counting up nearby NPCs.

Adjusted NPC scan range and possible NPC spawn range to use the same value. Changed this value to 90 OMTs radius (was previously 60 to scan, 120 to spawn)

Adjusted some debug screens to make it easier to tell how many NPCs exist (and how many are static)

#### Describe alternatives you've considered

The algorithm for random NPC spawns could probably use reconsidering, but that seemed like a much bigger process than what i did here.

I could have left it considering static/starting for the purposes of random NPCs, but that had the strange behavior that being near any static NPC settlements would more or less prohibit any random NPCs from spawning.

#### Testing

Compiled, played around with spawn settings (tried 100 density among other high values) to verify expected quantity of NPCs over time to verify the algorithm in an extreme case before adjusting to the "time in days" values, whereupon tested several sub-1-day values to get the expected hours between NPC spawns (following the scaling track)

Testing for really realz whether or not we're getting the expected amount of NPCs with our new default is a little trickier due to how much time is necessary to pass, and might require making test code i wasn't prepared to do just yet (but will if required).

#### Additional context

It's still possible to effectively disable random NPCs by setting the spawn time to `0.0`, at which point no random NPC will ever spawn.

the "lone wolf" flag can be added to scenarios to suppress the starting NPC, so modders can easily get the "always" or "never" effects by adjusting scenarios in mods.

if someone wants to not have static NPCs spawn at all, well, that's going to involve modding the locations where NPCs spawn. this will admittedly be quite a lot of work for anyone doing such a mod, but that effort barrier should also mean anyone who bothers will take the time to also adjust content so that the game is more fully playable with those NPCs missing.

Approximate expected average spawn times with the original 0.10 density value:
0 NPCs nearby, 42 days.
1 NPC nearby, 52 days.
5 NPCs nearby, 127 days.
10 NPCs nearby, 388 days.
20 NPCs nearby, 3,614 days (yes, almost ten years at this point, on average)

remember: this counted static NPCs as well (and the refugee center alone is over 20 NPCs).

Approximate expected average spawn times with the 1.0 spawn density value implied by the new default of 4 days spawn on average:
0 NPCs nearby, 4 days
1 NPC nearby, 5 days
5 NPCs nearby, 13 days
10 NPCs nearby, 39 days
20 NPCs nearby, 361 days

This is NOT counting static NPCs, but you still end up eventually hitting a "it takes a year to see another, probably" threshold. That point is just pushed back quite a bit between the new curve and the omission of static NPCs from the equation.

It's possible this is still too aggressively fast to spawn NPCs on the low end; i'm open to changes.

in order to determine if my stuff was working, i made some tweaks to the existing debug info UI to figure out how many NPCs i have.
This menu item:
image

has three screens, the third of which now is:
image

it always showed the locations of the 20 closest NPCs, (sorted by ostensible distance), but i added the third line, "Total NPCs within x OMTs", to help me figure out what was what.

### CODE REVIEW

**Comment by mlangsdorf** [comment]
I think the scan range and the spawn range should be the same.  Can we compromise and make them both 90?  That puts them inside an overmap buffer worth of OMTs.

Also, can we change the option to be "average number of days to create a random NPC", with a range from 3 to 60, and then calculate the actual spawn rate as 52 / value?  That way, it's much clearer to the player as to what is happening.

**Comment by esotericist** [comment]
> I think the scan range and the spawn range should be the same. Can we compromise and make them both 90? That puts them inside an overmap buffer worth of OMTs.

That makes sense to me, sure.

> Also, can we change the option to be "average number of days to create a random NPC", with a range from 3 to 60, and then calculate the actual spawn rate as 52 / value? That way, it's much clearer to the player as to what is happening.

Do we also try to communicate that the time goes up based on the number of nearby random NPCs? i wasn't really sure how to communicate details usefully without dropping a formula in the description.

**Comment by mlangsdorf** [comment]
I don't think it's worth explaining the nature of the exponential formula.  Days per first NPC is fine, and it's close enough for the first 10 NPCs anyway.

File: src/game.cpp [hunk 1]
```diff
@@ -11520,20 +11525,31 @@ void game::perhaps_add_random_npc()
...
         return;
     }
 
-    float density = get_option<float>( "NPC_DENSITY" );
-    static constexpr int density_search_radius = 60;
-    const float npc_num = overmap_buffer.get_npcs_near_player( density_search_radius ).size();
-    if( npc_num > 0.0 ) {
+    // spawn algorithm is a chance per hour, but the config is specified in average days
+    // actual chance per hour is (100 / 24 ) / days
+    static constexpr double days_to_rate_factor = ( 100 / 24 );
```
Comment (BevapDin) [inline src/game.cpp:11536]:
```suggestion
    static constexpr double days_to_rate_factor = 100.0 / 24;
```

**Comment by esotericist** [comment]
> Also, can we change the option to be "average number of days to create a random NPC", with a range from 3 to 60, and then calculate the actual spawn rate as 52 / value? That way, it's much clearer to the player as to what is happening.

I ended up allowing a range of 0 to 100 still. the 100 because i didn't see any particular need to restrict to less than a season. the 0 because that gives a useful sentinel value for preventing NPC spawns.

plus being able to set to a fraction of a day for a spawn timeframe is actually really useful for testing, it turns out :D

**Comment by esotericist** [comment]
Testing a fix for the clang-tidy complaint.

**Comment by esotericist** [comment]
I'm reasonably confident this is ready at this point.

**Comment by Broken27** [comment]
Why removing the starting NPC spawn option tho.

Sure you can still remove the starting NPC by modding the scenario, but that is an extra hassle that it's completely unecessary.

Starting NPC's affect the initial difficulty rather radically, and as such, they should always be optional.

**Comment by kevingranade** [comment]
Starting NPC's affect the initial difficulty rather radically, and as such, they should always be incorporated into the scenario.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/no-static-npc-toggle-option/25504/2

**Comment by int-ua** [comment]
So now there is absolutely no way to completely disable NPCs?

**Comment by esotericist** [comment]
> So now there is absolutely no way to completely disable NPCs?

I'm not sure why you would conclude that, given that I specify in this PR that 0 remains a valid value for random NPC speed, and mods can adjust static NPCs.

**Comment by int-ua** [comment]
sorry, TLDR

**Pull request merged**